Assert WebSession is not null

dukbong opened the pull request:
<!--
For Security Vulnerabilities, please use https://pivotal.io/security#reporting
-->
### Original Code:
The getAuthorizedClients method retrieves a map of OAuth2 authorized clients from a WebSession. In the original code, if the session is null, the method assigns a default empty HashMap. This code does not explicitly assert that WebSession cannot be null, potentially leading to silent failures or unexpected behavior.

Closes #14975

### Suggested Refactoring:
To ensure robustness and improve error handling, you can refactor the method by adding an assertion to prevent a null session. This approach uses Assert.notNull to raise an exception with a clear error message if the WebSession is null, helping developers quickly identify and fix the problem.

<!--
Before creating new features, we recommend creating an issue to discuss the feature. This ensures that everyone is on the same page before extensive work is done.

Thanks for contributing to Spring Security. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with gh-).
-->

dukbong commented:
> Hi @dukbong, thanks for the PR!
> 
> 
> 
> Just a heads up, I closed the original issue with [this explanation](https://github.com/spring-projects/spring-security/issues/14975#issuecomment-2090702647). I appreciate what you're doing in this change by asserting the session cannot be null. I'm wondering though if this change is absolutely necessary?
> 
> 
> 
> > This code does not explicitly assert that WebSession cannot be null, potentially leading to silent failures or unexpected behavior.
> 
> 
> 
> Can you explain the situation in which this would occur (other than the invalid configuration referred to in the original issue)?


Hi @sjohnr 

Initially, I planned to submit a pull request to address the issue, but as I reviewed the code, I found the structure to be questionable, prompting a change in direction.

The `getAuthorizedClient` function is written with the assumption that the session might be null. However, if the session is null, this leads to an exception in the `saveAuthorizedClient` function. To prevent this, I believe it's better to ensure the session isn't null either in the `getAuthorizedClient` function or in the `doOnSuccess` block.

Based on these observations, I created a pull request to address the issue.

- Your feedback is invaluable for my growth and learning.

sjohnr commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:
```diff
@@ -94,14 +94,11 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
+	    return authorizedClients != null ? authorizedClients : new HashMap<>();
 	}
 
+
```
Please remove this extra newline.

sjohnr commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:
Please update the copyright year in the header as per [contributing guidelines](https://github.com/spring-projects/spring-security/blob/main/CONTRIBUTING.adoc#submit-a-pull-request).

sjohnr commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:
```diff
@@ -94,14 +94,11 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
-		return authorizedClients;
+		Assert.notNull(session, "WebSession cannot be null. Ensure the session is properly initialized before accessing authorized clients.");
+	    Map<String, OAuth2AuthorizedClient> authorizedClients = session.getAttribute(this.sessionAttributeName);
+	    return authorizedClients != null ? authorizedClients : new HashMap<>();
```
You can add parentheses around the ternary condition to satisfy checkstyle and allow the build to pass.

sjohnr commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:
```diff
@@ -94,14 +94,11 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
-			authorizedClients = new HashMap<>();
-		}
-		return authorizedClients;
+		Assert.notNull(session, "WebSession cannot be null. Ensure the session is properly initialized before accessing authorized clients.");
```
I think the message can simply be `session cannot be null`.

sjohnr reviewed (changes requested):
Thanks for the PR @dukbong. In addition to the feedback inline below, can you please add a test that asserts an `IllegalArgumentException` is thrown when `WebSession` is null?

dukbong commented:
> Thanks for the PR @dukbong. In addition to the feedback inline below, can you please add a test that asserts an `IllegalArgumentException` is thrown when `WebSession` is null?

Thanks @sjohnr  for agreeing with my feedback.
I will make the changes and additions as per the comments below!

sjohnr commented on oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/server/WebSessionServerOAuth2AuthorizedClientRepository.java:
```diff
@@ -94,14 +94,10 @@ public Mono<Void> removeAuthorizedClient(String clientRegistrationId, Authentica
 		// @formatter:on
 	}
 
-	@SuppressWarnings("unchecked")
```
I don't think this should be removed. Can you restore this line?

sjohnr replied:
Apologies as you were correct to remove this. I'll re-remove it in a polish commit.

sjohnr reviewed (commented):
Thanks @dukbong. I added one minor comment below. Can you also please rebase your changes on main and squash commits? I will schedule this to be merged in the next minor so you won't see it merged for a few weeks.

dukbong commented:
> Thanks @dukbong. I added one minor comment below. Can you also please rebase your changes on main and squash commits? I will schedule this to be merged in the next minor so you won't see it merged for a few weeks.


Hello. @sjohnr 

As you suggested, I rebased onto the main branch and squashed the commits into one.

If there's anything I did wrong or need to correct, please let me know!
I'll fix it right away!!

sjohnr commented:
Thanks @dukbong, this is now merged into `5.8.x` as 1695d03b72d9d48f24abd8381f5f120b99a7f7a1. I did apply a polish commit with some adjustments because I didn't run the build to see checkstyle and other build output, so check out 
e34621ec2ced820accc121b053752387d053d9a5 and let me know if you have questions.
